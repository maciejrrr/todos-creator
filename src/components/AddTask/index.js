import React, { Component } from 'react';
import { uniqueId } from 'lodash';
import styled from 'styled-components';

import FormActions from '../FormActions';
import { metrics, colors } from '../../theme';

const CardFooter = styled.div`
  padding: ${metrics.smallMargin}px;
  background-color: ${colors.cardFooter};
  border-bottom-left-radius: ${metrics.borderRadius}px;
  border-bottom-right-radius: ${metrics.borderRadius}px;
`;

const CardForm = styled.form`
  padding: ${metrics.smallMargin}px;
`;

const FormSwitch = styled.a`
  color: ${colors.grayText};
  :hover {
    text-decoration: underline;
    color: ${colors.grayText};
  }
`;

const Textarea = styled.textarea`
  width: 100%;
  resize: none;
  border: 0;
  box-shadow: ${colors.boxShadow};
  padding: ${metrics.smallMargin}px;
  height: 5em;
  outline: 0;
  border-radius: ${metrics.borderRadius}px;
  vertical-align: top;
`;

class AddTask extends Component {
  state = {
    name: '',
    showForm: false,
  };

  handleSubmit = e => {
    e.preventDefault();
    const { addTask, cardId } = this.props;
    addTask({
      task: {
        id: uniqueId(),
        name: this.state.name,
        cardId,
      },
    });
    this.setState({ name: '' });
    this.textarea.focus();
  };

  handleShowForm = () => {
    this.setState({ showForm: true });
  };

  handleResetForm = () => {
    this.setState({ showForm: false, name: '' });
  };

  handleChange = e => {
    this.setState({ name: e.target.value });
  };

  render() {
    const { showForm, name } = this.state;
    let content = (
      <CardFooter>
        <FormSwitch onClick={this.handleShowForm}>Add a task...</FormSwitch>
      </CardFooter>
    );
    if (showForm) {
      content = (
        <CardForm onSubmit={this.handleSubmit}>
          <Textarea
            innerRef={textarea => (this.textarea = textarea)}
            autoFocus
            onChange={this.handleChange}
            value={name}
          />
          <FormActions
            focused={showForm}
            disabled={!name}
            text="Add"
            resetForm={this.handleResetForm}
          />
        </CardForm>
      );
    }

    return content;
  }
}

export default AddTask;
