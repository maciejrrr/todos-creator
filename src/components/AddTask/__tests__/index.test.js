import AddTask from '../index';

describe('AddTask component', () => {
  let props;
  beforeEach(() => {
    props = {
      addTask: jest.fn(),
      cardId: 1,
    };
  });

  describe('snapshots', () => {
    describe('initial state', () => {
      it('renders card footer with add task link', () => {
        const tree = renderer.create(<AddTask {...props} />).toJSON();
        expect(tree).toMatchSnapshot();
      });
    });

    describe('showForm state', () => {
      it('renders task form', () => {
        const component = renderer.create(<AddTask {...props} />);
        component.getInstance().handleShowForm();
        expect(component.toJSON()).toMatchSnapshot();
      });
    });
  });

  describe('handleSubmit', () => {
    let e;
    let preventDefaultMock;
    let focusMock;
    let name;
    let component;
    beforeEach(() => {
      preventDefaultMock = jest.fn();
      focusMock = jest.fn();
      e = { preventDefault: preventDefaultMock };
      name = 'valid name';
      component = shallow(<AddTask {...props} />);
      component.setState({ name, showForm: true });
      component.instance().textarea = { focus: focusMock };
      component.instance().handleSubmit(e);
    });

    it('calls preventDefault', () => {
      expect(preventDefaultMock).toHaveBeenCalledTimes(1);
    });

    it('calls addTask', () => {
      expect(props.addTask).toBeCalledWith({
        task: {
          id: 1,
          name,
        },
        cardId: 1,
      });
    });

    it('resets name state', () => {
      expect(component.state().name).toBe('');
    });

    it('calls focus on textarea ref', () => {
      expect(focusMock).toHaveBeenCalledTimes(1);
    });
  });

  describe('handleShowForm', () => {
    it('changes showForm state', () => {
      const component = shallow(<AddTask {...props} />);
      component.instance().handleShowForm();
      expect(component.state().showForm).toBe(true);
    });
  });

  describe('handleResetForm', () => {
    let component;
    beforeEach(() => {
      component = shallow(<AddTask {...props} />);
      component.setState({ showForm: true, name: 'test name' });
      component.instance().handleResetForm();
    });

    it('changes showForm state', () => {
      expect(component.state().showForm).toBe(false);
    });

    it('changes name state', () => {
      expect(component.state().name).toBe('');
    });
  });

  describe('handleChange', () => {
    it('changes name state', () => {
      const component = shallow(<AddTask {...props} />);
      const value = 'test value';
      const e = { target: { value } };
      component.instance().handleChange(e);
    });
  });
});
