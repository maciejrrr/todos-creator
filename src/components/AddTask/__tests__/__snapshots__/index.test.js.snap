// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddTask component snapshots initial state renders card footer with add task link 1`] = `
<div
  className="sc-bxivhb TDOcQ"
>
  <a
    className="sc-EHOje iCMDbo"
    onClick={[Function]}
  >
    Add a task...
  </a>
</div>
`;

exports[`AddTask component snapshots showForm state renders task form 1`] = `
<form
  className="sc-ifAKCX gYKwRG"
  onSubmit={[Function]}
>
  <textarea
    autoFocus={true}
    className="sc-bZQynM kKJhsz"
    onChange={[Function]}
    value=""
  />
  <div
    className="sc-bdVaJa cItXoG"
  >
    <button
      className="sc-bwzfXH hGZcQI"
      disabled={true}
    >
      Add
    </button>
    <i
      className="anticon anticon-close sc-htpNat oPmsX"
      onClick={[Function]}
    />
  </div>
</form>
`;
